lightkurve/correctors/pldcorrector.py

```diff
@@ -82,199 +116,509 @@ class PLDCorrector(object):
+        # This is an important step, as explained in Section 2 of Luger et al. (2016).
+        matrix = matrix / np.sum(matrix, axis=-1)[:, None]
+        # If we return matrix at this point, theano will raise a "dimension mismatch".
+        # The origin of this bug is not understood, but copying the matrix
+        # into a new one as shown below circumvents it:
+        result = np.empty((matrix.shape[0], matrix.shape[1]))
+        result[:, :] = matrix[:, :]
+
+        return result
+
+    def create_design_matrix(self, pld_order=1, n_pca_terms=10,
+                             include_column_of_ones=False, **kwargs):
+        """Returns a matrix designed to contain suitable regressors for the
+        systematics noise model.
+
+        The design matrix contains one row for each cadence (i.e. moment in time)
+        and one column for each regressor that we wish to use to predict the
+        systematic noise in a given cadence.
+
+        The columns (i.e. regressors) included in the design matrix are:
+        * One column for each pixel in the PLD aperture mask.  Each column
+          contains the flux values observed by that pixel over time.  This is
+          also known as the first order component.
+        * Columns derived from the products of all combinations of pixel values
+          in the aperture mask. However, rather than including a column for each
+          combination, we perform dimensionality reduction (PCA) and include a
+          smaller number of PCA terms, i.e. the number of columns is
+          n_pca_terms*(pld_order-1).  This is also known as the higher order
+          components.
+        * Optionally, a single column of ones for numerical stability.
+
+        Thus, the shape of the design matrix will be
+        (n_cadences, n_pld_mask_pixels + n_pca_terms*(pld_order-1) + include_column_of_ones)
 
-    def correct(self, aperture_mask=None, cadence_mask=None, gp_timescale=30,
-                use_gp=True, pld_order=2, n_pca_terms=10, pld_aperture_mask=None):
-        r"""Returns a PLD systematics-corrected LightCurve.
+        Returns
+        -------
+        design_matrix : 2D numpy array
+            See description above.
+        """
+        # We use an optional dependency for very fast PCA (fbpca), but if the
+        # import fails we will fall back on using the slower `np.linalg.svd`.
+        use_fbpca = True
+        try:
+            from fbpca import pca
+        except ImportError:
+            use_fbpca = False
+            log.warning("PLD systematics correction will run faster if the "
+                        "optional `fbpca` package is installed "
+                        "(`pip install fbpca`).")
+
+        matrix_sections = []  # list to hold the design matrix components
+        first_order_matrix = self.create_first_order_matrix()
+
+        # Input validation: n_pca_terms cannot be larger than the number of cadences
+        n_cadences = len(first_order_matrix)
```

Comment:
The user never specifies this number, but gets a warning if it's too small. This seems annoying. Can we instead just use

`n_cadences = np.min([len(first_order_matrix), n_cadences])`

Reply:
This check should actually ensure that `n_pca_terms` is not greater than the number of columns in the design matrix, so I've changed the name to `n_pixels`. This warning will only appear if the user passes in a very small PLD aperture mask (fewer than 4 pixels), which I don't think will be very common, and it prevents the code from breaking during PLD.

Do you think we should only give the warning if the user specifies `n_pca_terms`?